README.md
# dual: an R package for dual numbers 

### Authors <img src="man/figures/logo.svg" align="right" alt="dual logo" />
[Luca Sartore](mailto://drwolf85@gmail.com)

Maintainer: [Luca Sartore](mailto://drwolf85@gmail.com)

[![CRAN version](https://www.r-pkg.org/badges/version/dual)](https://cran.r-project.org/package=dual)
[![CRAN release](https://www.r-pkg.org/badges/ago/dual)](https://cran.r-project.org/package=dual)
[![GPLv3 license](https://img.shields.io/badge/License-GPLv3-yellow.svg)](https://perso.crans.org/besson/LICENSE.html)
[![CRAN RStudio mirror downloads](https://cranlogs.r-pkg.org/badges/dual)](https://cran.r-project.org/package=dual)
[![Total Downloads from CRAN RStudio mirror](https://cranlogs.r-pkg.org/badges/grand-total/dual?color=orange)](https://cran.r-project.org/package=dual)

## Features of the package
Dual numbers are mainly used to implement automatic differentiation. The **dual** package provides mathematical functions that are able to handle computations with dual numbers. The package is useful to calculate exact derivatives in R without providing self-coded functions.

For a complete list of exported functions, use `library(help = "dual")` once the **dual** package is installed (see the `inst/INSTALL.md` file for a detailed description of the setup process).

### Example
```R
library(dual)
x <- dual(f = 1.5, grad = c(1:0, 0))
y <- dual(f = 0.5, grad = c(0:1, 0))
z <- dual(f = 1.0, grad = c(0, 0:1))
exp(z - x) * sin(x)^y / x

a <- dual(f = 1.1, grad = c(1.2, 2.3, 3.4, 4.5, 5.6))
0.5 * a^2 - 0.1

lambertW <- function(x) {
  w0 <- 1
  w1 <- w0 - (w0*exp(w0)-x)/((w0+1)*exp(w0)-(w0+2)*(w0*exp(w0)-x)/(2*w0+2))
  while(abs(w1-w0) > 1e-15) {
    w0 <- w1
    w1 <- w0 - (w0*exp(w0)-x)/((w0+1)*exp(w0)-(w0+2)*(w0*exp(w0)-x)/(2*w0+2))
  }
  return(w1)
}
lambertW(dual(1, 1))
```

## References

Baydin, A. G., Pearlmutter, B. A., Radul, A. A., & Siskind, J. M. (2018). Automatic differentiation in machine learning: a survey. *Journal of Machine Learning Research*, **18**, 1-43.

Cheng, H. H. (1994). Programming with dual numbers and its applications in mechanisms design. *Engineering with Computers*, **10**(4), 212-229.

Kisil, V. V. (2007). Erlangen program at large-2: inventing a wheel. The parabolic one. *arXiv: 0707.4020*.


ChangeLog
2023-09-20 Luca Sartore <drwolf85@gmail.com>
    * DESCRIPTION (Version): New version is 0.0.5
    * DESCRIPTION (Date): New date is 2022-09-02
    * man/dual-package.Rd: Edited the version number, date and HTML5 
        features
    * dual-pkg.R: Edited roxygen `Title` and `Description` fields in
        the documentation of the package. The roxygen field
        `useDynLib` has been included too
    * src/dual.c: Included C interface with R to speed-up the
        computations required for derivatives of special functions
    * src/bessel.c: Included source code of particular Taylor series
        approximations used by the derivatives of Bessel functions
    * R/Arithmetic.R (line 217): Fixing signs in division method (bug
        reported by Philip Schuster)
    * R/Logic.R: Included methods for comparing dual numbers
    * README.md: Fixed another broken url

2022-08-31 Luca Sartore <drwolf85@gmail.com>
    * DESCRIPTION (Version): New version is 0.0.4
    * DESCRIPTION (Date): New date is 2022-08-31
    * DESCRIPTION (Description): Updated reference link to HTTPS
    * man/dual-package.Rd: Edited the version number, date and HTML5 
        conforming features

2019-12-14 Luca Sartore <drwolf85@gmail.com>
    * DESCRIPTION (Version): New version is 0.0.3
    * DESCRIPTION (Date): New date is 2019-12-14
    * DESCRIPTION (Description): Included a reference to expalin the
        methodology implemented in the package
    * README.md: Fixed a broken url
    * COPYING: Included the full text of the GPL-3 license

2019-11-20 Luca Sartore <drwolf85@gmail.com>
    * DESCRIPTION (Version): New version is 0.0.2
    * DESCRIPTION (Date): New date is 2019-11-20
    * DESCRIPTION (Description): The text has been rephrased to
        remove the name of the package

2019-11-19 Luca Sartore <drwolf85@gmail.com>
    * The dual package VERSION 0.0.1 has been created


R/dual-pkg.R
# @useDynLib dual, .registration = TRUE
#' @importFrom methods is new
#' @importFrom stats dnorm pchisq pnorm
#'
#' @name dual-package
#' 
#' @docType package
#' 
#' @encoding UTF-8
#' 
#' @title
#' \packageTitle{dual}
#' 
#' @description
#' \if{html}{\figure{logo.svg}{options: style='float: right;'}}
#' Automatic differentiation is achieved by using dual 
#' numbers without providing hand-coded gradient functions. 
#' The output value of a mathematical function is returned 
#' with the values of its exact first derivative (or 
#' gradient). For more details see Baydin, Pearlmutter, 
#' Radul, and Siskind (2018) 
#' \url{https://jmlr.org/papers/volume18/17-468/17-468.pdf}.
#' 
#' @details
#' \tabular{ll}{
#' Package: \tab dual\cr
#' Type: \tab Package\cr
#' Version: \tab 0.0.5\cr
#' Date: \tab 2023-10-02\cr
#' License: \tab GPL-3\cr
#' }
#' 
#' For a complete list of exported functions, use \code{library(help = "dual")}.
#'
#' @author
#' Luca Sartore \email{drwolf85@@gmail.com}
#'
#' Maintainer: Luca Sartore \email{drwolf85@@gmail.com}
#' 
#' @references
#' Baydin, A. G., Pearlmutter, B. A., Radul, A. A., & Siskind, J. M. (2018). Automatic differentiation in machine learning: a survey. \emph{Journal of Marchine Learning Research}, \bold{18}, 1-43.
#' 
#' Cheng, H. H. (1994). Programming with dual numbers and its applications in mechanisms design. \emph{Engineering with Computers}, \bold{10}(4), 212-229.
#'
#' @keywords dual autodiff differentiation numeric
#' 
#' @examples
#' library(dual)
#' 
#' # Initilizing variables of the function
#' x <- dual(f = 1.5, grad = c(1, 0, 0))
#' y <- dual(f = 0.5, grad = c(0, 1, 0))
#' z <- dual(f = 1.0, grad = c(0, 0, 1))
#' # Computing the function and its gradient
#' exp(z - x) * sin(x)^y / x
#' 
#' # General use for computations with dual numbers
#' a <- dual(1.1, grad = c(1.2, 2.3, 3.4, 4.5, 5.6))
#' 0.5 * a^2 - 0.1
#' 
#' # Johann Heinrich Lambert's W-function
#' lambertW <- function(x) {
#'   w0 <- 1
#'   w1 <- w0 - (w0*exp(w0)-x)/((w0+1)*exp(w0)-(w0+2)*(w0*exp(w0)-x)/(2*w0+2))
#'   while(abs(w1-w0) > 1e-15) {
#'     w0 <- w1
#'     w1 <- w0 - (w0*exp(w0)-x)/((w0+1)*exp(w0)-(w0+2)*(w0*exp(w0)-x)/(2*w0+2))
#'   }
#'   return(w1)
#' }
#' lambertW(dual(1, 1))
#' 
NULL
